File: YushanoBokenController.java
package YushanoBoken.Controller;

import YushanoBoken.Model.*;
import YushanoBoken.View.Option;

import java.util.ArrayList;
import java.util.Random;
import java.util.Scanner;

public class YushanoBokenController{
    private Scanner userInput = new Scanner(System.in);
    private int turn = 0;
    private Enemy enemy;
    private Hero hero;
    private Boss boss;
    private int level = 1;
    private ArrayList<BossName> bosses = new ArrayList<>();


    public int startGame(Hero hero) {
        this.hero = hero;
        generateEnemiesOrBoss();
        presentEnemyOrBoss();
        bosses.clear();
        return turn = 0;
    }

    public int HeroTurn(Option option){
        if(hero.isDisable()){
            return turn = 1;
        }
        if(option == Option.Attack){
            if (enemy != null) {
                enemy.recieveDamage(hero.getAttack());
                if (enemy.isDead()){
                    defeatEnemy(enemy);
                    return turn = 0;
                }
            } else if (boss != null) {
                boss.recieveDamage(hero.getAttack());
                if (boss.isDead()){
                    defeatBoss(boss);
                    return turn = 0;
                }
            }
        } else if (option == Option.Use_Item){
            useItem();
        } else if (option == Option.Magic){
            useMagic();
        }
        return turn = 1;
    }

    private void defeatEnemy(Enemy enemy) {
        System.out.println("+++" + enemy.getName() + " iS DEAD +++ \n");
        Item item = enemy.dropLoot();
        hero.addItem(item);
        System.out.println("===You receive " + item + "===");
        hero.levelUp(enemy.getExp());
        System.out.println("===You got " + enemy.getExp() + " exp===");
        this.enemy = null;
        generateEnemiesOrBoss();
        presentEnemyOrBoss();
    }

    private void defeatBoss(Boss boss) {
        Item item = boss.dropLoot();
        hero.addItem(item);
        System.out.println("======You receive " + item + "=====");
        hero.levelUp(boss.getExp());
        System.out.println("======You got " + boss.getExp() + " exp ======");
        System.out.println("CONGRATULATIONS \nYou have defeated " + boss.getName() + "\n");
        this.boss = null;
        generateEnemiesOrBoss();
        presentEnemyOrBoss();
    }

    private void useItem() {
        boolean isInventoryEmpty = hero.displayInventory();
        if (isInventoryEmpty) {
            System.out.println("Chose an item # \n");
            Item item = choseAOption();
            hero.useItem(item);
        } else {
            System.out.println("You don't have items in your inventory \n");
        }
    }

    private void useMagic() {
        boolean isMagicEmpty = hero.displayMagic();
        if (!isMagicEmpty) {
            System.out.println("You have no magic available.");
            return;
        }
        System.out.println("Select a magic: ");
        int option = userInput.nextInt() - 1;
        if (option >= 0 && option < hero.getMagic().size()) {
            Magic magia = hero.getMagic().get(option);
            if (hero.getMp() >= magia.getMpCost()) {
                hero.setMp(hero.getMp() - magia.getMpCost());
                System.out.println(hero.getName() + " use " + magia.getType());
                if(magia.getType() == MagicType.Cure || magia.getType() == MagicType.Aero){
                    hero.setEffect(magia.getEffectType());
                }else{
                    if (enemy != null) {
                        enemy.recieveDamage(magia.getEffect());
                        enemy.setEffect(magia.getEffectType());
                        if (enemy.isDead()) {
                            defeatEnemy(enemy);
                        }
                    } else if (boss != null) {
                        boss.recieveDamage(magia.getEffect());
                        boss.setEffect(magia.getEffectType());
                        if (boss.isDead()) {
                            defeatBoss(boss);
                        }
                    }
                }
            } else {
                System.out.println("You don't have enough MP.");
            }
        } else {
            System.out.println("Invalid option.");
        }
    }

    public Item choseAOption(){
        String option = userInput.nextLine();
        return chosenOption(option);
    }

    public Item chosenOption(String option) {
        int optionIndex = 0;
        try {
            optionIndex = Integer.parseInt(option);
        } catch (NumberFormatException e) {
            System.out.println("Please enter a valid number!");
            return choseAOption();
        }
        Inventory heroInventory = hero.getInventory();
        ArrayList<Item> items = heroInventory.getItems();
        if (optionIndex < 1 || optionIndex > items.size()) {
            System.out.println("Please enter a valid option!");
            System.out.println("Enter a number between 1 and " + items.size() + ".");
            return choseAOption();
        }

        return items.get(optionIndex - 1);
    }

    public int EnemiesTurn(){
        if (enemy != null) {
            enemy.applyStatusEffect();
            if (enemy.isDisable()){
                updateEnemyAndBossHealth();
                return 0;
            }
            updateEnemyAndBossHealth();
            if(enemy.isDead()){
                defeatEnemy(enemy);
                return 0;
            }
            hero.recieveDamage(enemy.getAttack());
            System.out.println("++++YOU HAVE RECEIVED " + hero.getDamageAfterDefense() + " DAMAGE+++++");
            return turn = 0;
        } else if (boss != null) {
            boss.applyStatusEffect();
            if (boss.isDisable()){
                updateEnemyAndBossHealth();
                return 0;
            }
            updateEnemyAndBossHealth();
            if(boss.isDead()){
                defeatBoss(boss);
                return 0;
            }
            Random rand = new Random();
            int useSpecialAbility = rand.nextInt(3) + 1;
            if (useSpecialAbility == 3) {
                boss.useSpecialAbility(hero);
                System.out.println("++++++" + boss.getName() + " USE " + boss.getAbilty() + " SPECIAL+++++");
            } else {
                hero.recieveDamage(boss.getAttack());
                System.out.println("+++++YOU HAVE RECEIVED " + hero.getDamageAfterDefense() + " DAMAGE++++++");
            }
        }
        return turn = 0;
    }

    public void generateEnemiesOrBoss() {
        if (enemy == null || boss == null || enemy.isDead() || boss.isDead()) {
            Random rand = new Random();
            int generateEnemiesOrBoss = rand.nextInt(5) + 1;
            if (hero.getLevel() >= 10 && generateEnemiesOrBoss >= 4) {
                generateBossBasedOnLevel();
            } else {
                generateRandomEnemy();
            }
        }
    }

    private void generateBossBasedOnLevel() {
        int heroLevel = hero.getLevel();
        if (heroLevel >= 10 && heroLevel <= 20) {
            generateBoss(BossName.SOLARIAS_IGNIS);
        } else if (heroLevel >= 21 && heroLevel <= 30) {
            generateBoss(BossName.VACORIUS_NEXIS);
        } else if (heroLevel >= 31 && heroLevel <= 40) {
            generateBoss(BossName.SOLARIAS_ULTIMA);
        } else if (heroLevel >= 50 && heroLevel <= 100) {
            generateBoss(BossName.SHIN_GODZILLA);
        }
    }

    private void generateBoss(BossName bossName) {
        if (bosses.contains(bossName)) {
            generateRandomEnemy();
        } else {
            boss = new Boss(bossName);
            bosses.add(bossName);
        }
    }

    private void generateRandomEnemy() {
        Random rand = new Random();
        EnemyType[] enemyTypes = EnemyType.values();
        int enemies = rand.nextInt(enemyTypes.length);
        this.enemy = new Enemy(enemyTypes[enemies]);
        if (hero.getLevel() >= 6) {
            enemy.levelUp(getLevel());
            setLevel(getLevel() + 1);
        }
    }

    public void setLevel(int level){
        this.level = level;
    }

    public int getLevel(){
        return level;
    }

    public void updateEnemyAndBossHealth(){
        StringBuilder hp = new StringBuilder();
        if (enemy != null) {
            hp.append("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\n");
            hp.append("        ⚔️ ENEMY IN COMBAT ⚔️\n");
            hp.append("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\n");
            hp.append("👹 Name: " + enemy.getName() + "\n");
            hp.append("❤️ HP: " + enemy.getHp() + "\n");
            if(enemy.getEffectDisplay() != null){
                hp.append("✨ Status: " + enemy.getEffectDisplay() +"\n");
            }
            hp.append("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\n");
        } else if (boss != null) {
            hp.append("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\n");
            hp.append("       ⚠️ BOSS IN COMBAT ⚠️\n");
            hp.append("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\n");
            hp.append("👿 Name: " + boss.getName() + "\n");
            hp.append("❤️ HP: " + boss.getHp() + "\n");
            if(boss.getEffectDisplay() != null){
                hp.append("✨ Status: " + boss.getEffectDisplay() +"\n");
            }
            hp.append("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\n");
        }
        System.out.println(hp);
    }

    public void presentEnemyOrBoss(){
        StringBuilder displayEnemyInfo = new StringBuilder();
        if (enemy != null) {
            displayEnemyInfo.append("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\n");
            displayEnemyInfo.append("         AN ENEMY HAS APPEARED           \n");
            displayEnemyInfo.append("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\n");
            displayEnemyInfo.append("        ⚔️ ENEMY IN COMBAT ⚔️\n");
            displayEnemyInfo.append("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\n");
            displayEnemyInfo.append("👹 Name: " + enemy.getName() + "\n");
            displayEnemyInfo.append("❤️ HP: " + enemy.getHp() + "\n");
            displayEnemyInfo.append("⚔️ Attack: " + enemy.getAttack() + "\n");
            displayEnemyInfo.append("🛡️ Defense: " + enemy.getDefense() + "\n");
            displayEnemyInfo.append("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\n");
        } else if (boss != null){
            displayEnemyInfo.append("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\n");
            displayEnemyInfo.append("       ⚠️⚠️⚠️⚠️DANGER⚠️⚠️⚠️⚠️          \n");
            displayEnemyInfo.append("         AN BOSS HAS APPEARED            \n");
            displayEnemyInfo.append("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\n");
            displayEnemyInfo.append("       ⚠️ BOSS IN COMBAT ⚠️\n");
            displayEnemyInfo.append("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\n");
            displayEnemyInfo.append("👿 Name: " + boss.getName() + "\n");
            displayEnemyInfo.append("❤️ HP: " + boss.getHp() + "\n");
            displayEnemyInfo.append("⚔️ Attack: " + boss.getAttack() + "\n");
            displayEnemyInfo.append("🛡️ Defense: " + boss.getDefense() + "\n");
            displayEnemyInfo.append("⚡ Special Ability: " + boss.getAbilty() + "\n");
            displayEnemyInfo.append("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\n");
        }
        System.out.println(displayEnemyInfo);
    }

}

File: RPG/YushanoBokenController.cpp
#include "YushanoBokenController.h"
#include <iostream>

YushanoBokenController::YushanoBokenController()
    : turn(0), enemy(nullptr), hero(nullptr), boss(nullptr), level(1) {
    std::srand(std::time(nullptr));
}

int YushanoBokenController::startGame(Hero* hero) {
    this->hero = hero;
    generateEnemiesOrBoss();
    presentEnemyOrBoss();
    bosses.clear();
    return turn = 0;
}

int YushanoBokenController::HeroTurn(Option option) {
    if (hero->IsDisabled()) {
        return turn = 1;
    }

    if (option == Option::Attack) {
        if (enemy != nullptr) {
            enemy->ReceiveDamage(hero->GetAttack());
            if (enemy->IsDead()) {
                defeatEnemy(enemy);
                return turn = 0;
            }
        }
        else if (boss != nullptr) {
            boss->ReceiveDamage(hero->GetAttack());
            if (boss->IsDead()) {
                defeatBoss(boss);
                return turn = 0;
            }
        }
    }
    else if (option == Option::Use_Item) {
        useItem();
    }
    else if (option == Option::Magic) {
        useMagic();
    }
    return turn = 1;
}

void YushanoBokenController::defeatEnemy(Enemy* enemy) {
    std::cout << "+++" << enemy->GetName() << " is DEAD +++\n";
    Item* item2 = enemy->DropLoot();
    hero->AddItem(item2);
    std::cout << "===You receive " << item2->ToString() << "===\n";
    hero->LevelUp(enemy->GetExp());
    std::cout << "===You got " << enemy->GetExp() << " exp===\n";
    this->enemy = nullptr;
    generateEnemiesOrBoss();
    presentEnemyOrBoss();
}

void YushanoBokenController::defeatBoss(Boss* boss) {
    Item* item = boss->DropLoot();
    hero->AddItem(item);
    std::cout << "=====You receive " << item->ToString() << "=====\n";
    hero->LevelUp(boss->GetExp());
    std::cout << "=====You got " << boss->GetExp() << " exp =====\n";
    std::cout << "CONGRATULATIONS\nYou have defeated " << boss->GetName() << "\n";
    this->boss = nullptr;
    generateEnemiesOrBoss();
    presentEnemyOrBoss();
}

void YushanoBokenController::useItem() {
    bool isInventoryEmpty = hero->DisplayInventory();
    if (!isInventoryEmpty) {
        std::cout << "Choose an item #\n";
        int option;
        std::cin >> option;
        Item* item = chosenOption(option);
        hero->UseItem(item);
    }
    else {
        std::cout << "You don't have items in your inventory\n";
    }
}

void YushanoBokenController::useMagic() {
    bool isMagicEmpty = hero->DisplayMagics();
    if (isMagicEmpty) {
        std::cout << "You have no magic available.\n";
        return;
    }

    std::cout << "Select a magic: ";
    int option;
    std::cin >> option;
    option--; // Convert to 0-based index

    if (option >= 0 && option < hero->GetMagicCount()) {
        Magic* magic = hero->GetMagics(option);
        if (hero->GetMp() >= magic->GetMpCost()) {
            hero->SetMp(hero->GetMp() - magic->GetMpCost());
            std::cout << hero->GetName() << " use " << magic->GetTypeString() << "\n";

            if (magic->GetType() == MagicType::CURE || magic->GetType() == MagicType::AERO) {
                hero->SetEffect(magic->GetEffectType());
            }
            else {
                if (enemy != nullptr) {
                    enemy->ReceiveDamage(magic->GetEffect());
                    enemy->SetEffect(magic->GetEffectType());
                    if (enemy->IsDead()) {
                        defeatEnemy(enemy);
                    }
                }
                else if (boss != nullptr) {
                    boss->ReceiveDamage(magic->GetEffect());
                    boss->SetEffect(magic->GetEffectType());
                    if (boss->IsDead()) {
                        defeatBoss(boss);
                    }
                }
            }
        }
        else {
            std::cout << "You don't have enough MP.\n";
        }
    }
    else {
        std::cout << "Invalid option.\n";
    }
}

int YushanoBokenController::EnemiesTurn() {
    if (enemy != nullptr) {
        enemy->ApplyStatusEffect();
        if (enemy->IsDisable()) {
            updateEnemyAndBossHealth();
            return 0;
        }

        updateEnemyAndBossHealth();
        if (enemy->IsDead()) {
            defeatEnemy(enemy);
            return 0;
        }

        hero->ReceiveDamage(enemy->GetAttack());
        std::cout << "++++YOU HAVE RECEIVED " << hero->GetDamageAfterDefense() << " DAMAGE+++++\n";
        return turn = 0;
    }
    else if (boss != nullptr) {
        boss->ApplyStatusEffect();
        if (boss->isDisable()) {
            updateEnemyAndBossHealth();
            return 0;
        }

        updateEnemyAndBossHealth();
        if (boss->IsDead()) {
            defeatBoss(boss);
            return 0;
        }

        int useSpecialAbility = std::rand() % 3 + 1;
        if (useSpecialAbility == 3) {
            boss->UseSpecialAbility(hero);
            std::cout << "++++++" << boss->GetName() << " USE " << boss->getAbility() << " SPECIAL+++++\n";
        }
        else {
            hero->ReceiveDamage(boss->GetAttack());
            std::cout << "+++++YOU HAVE RECEIVED " << hero->GetDamageAfterDefense() << " DAMAGE++++++\n";
        }
    }
    return turn = 0;
}

void YushanoBokenController::generateEnemiesOrBoss() {
    if (enemy == nullptr && boss == nullptr) {
        int generateEnemiesOrBoss = std::rand() % 5 + 1;
        if (hero->GetLevel() >= 10 && generateEnemiesOrBoss >= 4) {
            generateBossBasedOnLevel();
        }
        else {
            generateRandomEnemy();
        }
    }
}

void YushanoBokenController::generateBossBasedOnLevel() {
    int heroLevel = hero->GetLevel();
    if (heroLevel >= 10 && heroLevel <= 20) {
        generateBoss(BossName::SOLARIAS_IGNIS);
    }
    else if (heroLevel >= 21 && heroLevel <= 30) {
        generateBoss(BossName::VACORIUS_NEXIS);
    }
    else if (heroLevel >= 31 && heroLevel <= 40) {
        generateBoss(BossName::SOLARIAS_ULTIMA);
    }
    else if (heroLevel >= 50 && heroLevel <= 100) {
        generateBoss(BossName::SHIN_GODZILLA);
    }
}

void YushanoBokenController::generateBoss(BossName bossName) {
    if (std::find(bosses.begin(), bosses.end(), bossName) != bosses.end()) {
        generateRandomEnemy();
    }
    else {
        boss = new Boss(bossName);
        bosses.push_back(bossName);
        AddMessage("A terrifying boss has appeared: " + boss->GetName() + "!");
    }
}

void YushanoBokenController::generateRandomEnemy() {
    EnemyType enemyTypes[] = { /* Your enemy types here */ };
    int randomIndex = std::rand() % (sizeof(enemyTypes) / sizeof(enemyTypes[0]));
    enemy = new Enemy(enemyTypes[randomIndex]);

    if (hero->GetLevel() >= 6) {
        enemy->LevelUp(level);
        level++;
    }

    AddMessage("An enemy appears: " + enemy->GetName() + "!");
}

Item* YushanoBokenController::chosenOption(int option) {
    try {
        if (option < 1 || option > hero->GetInventory().GetItems().size()) {
            throw std::out_of_range("Invalid option");
        }
        return hero->GetInventory().GetItems()[option - 1];
    }
    catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        std::cerr << "Please enter a number between 1 and "
            << hero->GetInventory().GetItems().size() << std::endl;
        return nullptr;
    }
}

void YushanoBokenController::updateEnemyAndBossHealth() {
    std::string status;
    if (enemy != nullptr) {
        status = "ENEMY: " + enemy->GetName() +
            "\nHP: " + std::to_string(enemy->GetHp()) +
            "/" + std::to_string(enemy->GetMaxHp());
        if (!enemy->GetEffectDisplay().empty()) {
            status += "\nStatus: " + enemy->GetEffectDisplay();
        }
        AddMessage(status);
    }
    else if (boss != nullptr) {
        status = "BOSS: " + boss->GetName() +
            "\nHP: " + std::to_string(boss->GetHp()) +
            "/" + std::to_string(boss->GetMaxHp());
        if (!boss->GetEffectDisplay().empty()) {
            status += "\nStatus: " + boss->GetEffectDisplay();
        }
        AddMessage(status);
    }
}

void YushanoBokenController::presentEnemyOrBoss() {
    if (enemy != nullptr) {
        std::string info = "ENEMY ENCOUNTER!\n" +
            enemy->GetName() +
            "\nATK: " + std::to_string(enemy->GetAttack()) +
            "\nDEF: " + std::to_string(enemy->GetDefense());
        AddMessage(info);
    }
    else if (boss != nullptr) {
        std::string info = "BOSS BATTLE!\n" +
            boss->GetName() +
            "\nATK: " + std::to_string(boss->GetAttack()) +
            "\nDEF: " + std::to_string(boss->GetDefense()) +
            "\nSpecial: " + boss->GetAbility();
        AddMessage(info);
    }
}

void YushanoBokenController::AddMessage(const std::string& message) {
    messages.push_back(message);
    // Keep only the last 5 messages
    if (messages.size() > 5) {
        messages.erase(messages.begin());
    }
    messageTimer = 5.0f; // Display for 5 seconds
}

int YushanoBokenController::getLevel() const {
    return level;
}

void YushanoBokenController::setLevel(int level) {
    this->level = level;
}
